include/fastscapelib/profile_grid.hpp

```diff
@@ -0,0 +1,283 @@
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <cstddef>
+#include <cstdint>
+#include <type_traits>
+#include <map>
+#include <vector>
+
+
+namespace fastscapelib
+{
+
+    /**
+     * Status at grid boundary nodes.
+     */
+    class profile_boundary_status: public boundary_status
+    {
+    public:
+
+        node_status left = node_status::core;    /**< Status at left edge/border node(s) */
+        node_status right = node_status::core;   /**< Status at right edge/border node(s) */
+
+        profile_boundary_status(const node_status status);
+        profile_boundary_status(const node_status left_status, const node_status right_status);
+        profile_boundary_status(const std::array<node_status, 2>& status);
+
+        bool is_horizontal_looped();
+
+    protected:
+
+        void check_looped_symmetrical();
+    };
+
+    /**
+     * @name Constructors
+     */
+    //@{
+    /**
+     * Set the same status for all boundary nodes.
+     */
+    inline profile_boundary_status::profile_boundary_status(node_status status)
+        : left(status), right(status)
+    {
+        check_looped_symmetrical();
+    }
+
+    /**
+     * Set the same status for all boundary nodes.
+     */
+    inline profile_boundary_status::profile_boundary_status(node_status left_status, node_status right_status)
+        : left(left_status), right(right_status)
+    {
+        check_looped_symmetrical();
+    }
+
+    /**
+     * Set status at the left and right edge nodes of a profile grid.
+     */
+    inline profile_boundary_status::profile_boundary_status(const std::array<node_status, 2>& status)
+        : left(status[0]), right(status[1])
+    {
+        check_looped_symmetrical();
+    }
+    //@}
+
+    /**
+     * Return true if periodic (looped) conditions are set for ``left`` and ``right``.
+     */
+    inline bool profile_boundary_status::is_horizontal_looped()
+    {
+        return is_looped(left) && is_looped(right);
+    }
+
+    inline void profile_boundary_status::check_looped_symmetrical()
+    {
+        if (is_looped(left) ^ is_looped(right))
+        {
+            throw std::invalid_argument("looped boundaries are not symmetrical");
+        }
+    }
+
+    //*******************
+    //* Profile grid (1D)
+    //*******************
+
+    template <class XT>
+    class profile_grid_xt;
+
+    template <class XT>
+    struct grid_inner_types<profile_grid_xt<XT>>
+    {
+        using xt_selector = XT;
+
+        struct xt_ndims
+        {
+            static constexpr std::size_t value = 1;
+        };
+
+        using xt_type = xt_container_t<xt_selector, int, xt_ndims::value>;
+
+        using size_type = typename xt_type::size_type;
+        using shape_type = typename xt_type::shape_type;
+
+        struct is_structured
+        {
+            static constexpr bool value = true;
+        };
+
+        struct is_uniform
+        {
+            static constexpr bool value = true;
```

Comment:
I think that it will eventually be useful to support non-uniform profile grids (this could be done later). What's the best approach here?

1. add another grid class?
2. is there a way to allow `spacing_type` be either `double` or `xt_container_t<xt_selector, double, xt_ndims::value>`, and statically compute `is_uniform` depending on the given `spacing_type`? Could we do this without exposing another template parameter to `profile_grid_xt` (i.e., infer `spacing_type` from the type given in the constructor)?

Reply:
As a first guess, I would say that a dedicated class will offer better performance because the caches of distances will be optimized for the uniform cases.

I will open an issue to track that idea!